### PR TITLE
Cloud upload implementation

### PR DESCRIPTION
**Pull request by Prajwal17Tunerlabs**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: controllers/v1/cloud-services/aws.js [hunk 1]
```diff
@@ -0,0 +1,275 @@
+/**
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/cloud-services/aws.js:1]: @Prajwal17Tunerlabs We don't need this file. remove it. Check if we can remove any other code, which is only used by this file

File: controllers/v1/cloud-services/files.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+      * @name upload
+      * @param  {Request}  req  request body.
+      * @returns {JSON} Response with status .
+      * @apiParamExample {json} Response:
+     * {
+            "status": 200
+       }
+     */
+
+  async upload(req) {
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/cloud-services/files.js:178]: @Prajwal17Tunerlabs this api is tested? Add comments to the code to increase readability

File: envVariables.js [hunk 3]
```diff
@@ -61,127 +61,165 @@ let enviromentVariables = {
     "optional" : false,
     "default" : "ON"
   },
-  "CLOUD_STORAGE" : {
-    "message" : "Enable/Disable cloud services",
+  // "CLOUD_STORAGE" : {
```
Comment (VISHNUDAS-tunerlabs) [inline envVariables.js:64]: @Prajwal17Tunerlabs remove unused variables from the list

File: envVariables.js [hunk 4]
```diff
@@ -61,127 +61,165 @@ let enviromentVariables = {
...
-  }, 
-  "AWS_BUCKET_REGION" : {
-    "message" : "Required Aws bucket region",
-    "optional" : true,
-    "requiredIf" : {
-      "key": "CLOUD_STORAGE",
-      "operator": "EQUALS",
-      "value" : "AWS"
-    }
+  "CLOUD_STORAGE_ACCOUNTNAME" : {
```
Comment (VISHNUDAS-tunerlabs) [inline envVariables.js:209]: provider: process.env.CLOUD_STORAGE_PROVIDER,
      identity: process.env.CLOUD_STORAGE_ACCOUNTNAME,
      credential: process.env.CLOUD_STORAGE_SECRET,
      publicContainer: process.env.CLOUD_STORAGE_BUCKETNAME,
      @Prajwal17Tunerlabs  Above mentioned are mandatory. remove if anything else is specified

File: generics/constants/common.js [hunk 5]
```diff
@@ -69,5 +69,13 @@ module.exports = {
     "CREATE_SURVEY": "/create-survey/",
     "SUBMISSION_STATUS_COMPLETED" : "completed",
     "EXPIRED": "expired",
-    "STARTED":"started"
+    "STARTED":"started",
+    "AWS_SERVICE": "AWS",
+    "NO_OF_MINUTES": 30,
+    "NO_OF_EXPIRY_TIME": 60,
```
Comment (VISHNUDAS-tunerlabs) [inline generics/constants/common.js:75]: @Prajwal17Tunerlabs We need to add two more env Variables
PRESIGNED_URL_EXPIRY_IN_SECONDS (in seconds, when we create signed url , by default pass it )
DOWNLOADABLE_URL_EXPIRY_IN_SECONDS (in seconds, when ever downloadable url created for private bucket pass this value as expiry )

update envVariables.js
sample : 
DOWNLOAD_URL_EXPIRATION_DURATION: {
		message: 'Required downloadable url expiration time',
		optional: true,
		default: 300,
	},

File: envVariables.js [hunk 6]
```diff
@@ -61,127 +61,165 @@ let enviromentVariables = {
...
+  "CLOUD_STORAGE_BUCKET_TYPE" : {
+    "message" : "Require client storage bucket type",
+    "optional" : false
+  },
+  "PUBLIC_ASSET_BUCKETNAME" : {
+    "message" : "Require public asset bucket name",
+    "optional" : false
+  },
+
+
```
Comment (VISHNUDAS-tunerlabs) [inline envVariables.js:222]: Add this env also
ALLOWED_HOST: {
		message: 'Required CORS allowed host',
		optional: true,
		default: '*',
	},

File: module/cloud-services/files/helper.js [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+   * @name preSignedUrls
+   * @param {Array} payloadData       - payload for files data.
+   * @param {String} referenceType    - reference type
+   * @param {String} userId           - Logged in user id.
+   * @param {String} templateId       - certificateTemplateId.
+   * @param {Boolean} serviceUpload     - serive Upload  {true/false}
+   * @returns {Array}                 - consists of all signed urls files.
+   */
+
+  static preSignedUrls(payloadData, userId = "") {
```
Comment (VISHNUDAS-tunerlabs) [inline module/cloud-services/files/helper.js:36]: @Prajwal17Tunerlabs Add proper comments to increase readability of the code

File: module/files/helper.js [hunk 8]
```diff
@@ -1,126 +1,403 @@
...
+  static getDownloadableUrl(
+    filePath,
+    bucketName,
+    storageName = "",
+    expireIn = ""
+  ) {
+    return new Promise(async (resolve, reject) => {
+      try {
+        let noOfMinutes = CONSTANTS.common.NO_OF_MINUTES;
+        let linkExpireTime = CONSTANTS.common.NO_OF_EXPIRY_TIME * noOfMinutes;
```
Comment (VISHNUDAS-tunerlabs) [inline module/files/helper.js:41]: @Prajwal17Tunerlabs make use of env variables

File: module/files/helper.js [hunk 9]
```diff
@@ -1,126 +1,403 @@
...
+        // Override linkExpireTime if provided explicitly.
+        if (expireIn !== "") {
+          linkExpireTime = expireIn;
+        }
+
+        if (Array.isArray(filePath) && filePath.length > 0) {
+          let result = [];
 
-	/**
+          await Promise.all(
```
Comment (VISHNUDAS-tunerlabs) [inline module/files/helper.js:55]: @Prajwal17Tunerlabs Please Add proper comments in the files for increasing the readability of the code

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
Review during development, not on final PR. 29-April-2024 : 12:50 PM

File: module/files/helper.js (hunk 8 above)
Comment (Prajwal17Tunerlabs) [reply module/files/helper.js:41]: using env variable instead of common.js

File: envVariables.js (hunk 6 above)
Comment (Prajwal17Tunerlabs) [reply envVariables.js:222]: added the above env variable

File: controllers/v1/cloud-services/aws.js (hunk 1 above)
Comment (Prajwal17Tunerlabs) [reply controllers/v1/cloud-services/aws.js:1]: removed aws.js file

File: envVariables.js (hunk 3 above)
Comment (Prajwal17Tunerlabs) [reply envVariables.js:64]: removed unsued variables

File: module/files/helper.js (hunk 9 above)
Comment (Prajwal17Tunerlabs) [reply module/files/helper.js:55]: added a comment

File: module/cloud-services/files/helper.js (hunk 7 above)
Comment (Prajwal17Tunerlabs) [reply module/cloud-services/files/helper.js:36]: added comments to increase readability of the code

File: generics/constants/common.js (hunk 5 above)
Comment (Prajwal17Tunerlabs) [reply generics/constants/common.js:75]: have added the specified env variables in envVaribales.js and .env files

File: controllers/v1/cloud-services/files.js (hunk 2 above)
Comment (Prajwal17Tunerlabs) [reply controllers/v1/cloud-services/files.js]: this api is removed since we are not using it anymore

File: controllers/v1/solutions.js [hunk 10]
```diff
@@ -281,7 +281,11 @@ module.exports = class Solutions extends Abstract {
     /**
     * @api {post} /project/v1/solutions/addRolesInScope/:solutionId Add roles in solutions
     * @apiVersion 1.0.0
+<<<<<<< HEAD
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/solutions.js:284]: @Prajwal17Tunerlabs remove these. Make sure merge doesn't cause any code change

File: generics/constants/common.js [hunk 11]
```diff
@@ -69,5 +69,11 @@ module.exports = {
...
     "SUBMISSION_STATUS_COMPLETED" : "completed",
     "EXPIRED": "expired",
-    "STARTED":"started"
+    "STARTED":"started",
+    "AWS_SERVICE": "AWS",
+    "WRITE_PERMISSION": "w",
+    "READ_PERMISSION": "r",
+    "PUBLIC" : "public",
+    "PRIVATE" : "private",
+    "UPLOAD_FILE":"project/cloud-services/elevate-project/v1/files/upload"
```
Comment (VISHNUDAS-tunerlabs) [inline generics/constants/common.js:78]: @Prajwal17Tunerlabs is this used any where?

File: module/cloud-services/files/helper.js [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+					let payloadId = payloadIds[pointerToPayload]
+					// Generate unique folderPath to all the file names in payloadData
+					let folderPath = 'project/' + payloadId + '/' + userId + '/' + UTILS.generateUniqueId() + '/'
+					// Call preSignedUrls helper file to get the signedUrl
+					let imagePayload = await filesHelpers.preSignedUrls(
+						payloadData[payloadId].files,
+						bucketName,
+						cloudStorage,
+						folderPath,
+						typeof process.env.PRESIGNED_URL_EXPIRY_IN_SECONDS === 'string'
```
Comment (VISHNUDAS-tunerlabs) [inline module/cloud-services/files/helper.js:54]: @Prajwal17Tunerlabs do we need this check? we can convert to int even if it is a number. avoid the check always convert to int. make sure to test giving as string and link is expiring after the expiry time

File: module/cloud-services/files/helper.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+		return new Promise(async (resolve, reject) => {
+			try {
+				// if bucktType is private call preSignedUrls function with read permission
+				if (bucktType === CONSTANTS.common.PRIVATE) {
+					let downloadableUrl = await filesHelpers.preSignedUrls(
+						payloadData,
+						bucketName,
+						cloudStorage,
+						'',
+						typeof process.env.DOWNLOADABLE_URL_EXPIRY_IN_SECONDS === 'string'
```
Comment (VISHNUDAS-tunerlabs) [inline module/cloud-services/files/helper.js:103]: @Prajwal17Tunerlabs here also we can convert to int without check

File: module/cloud-services/files/helper.js [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+						message: CONSTANTS.apiResponses.CLOUD_SERVICE_SUCCESS_MESSAGE,
+						result: downloadableUrl.result,
+					})
+				}
+				// if bucketType is public go with the normal flow of calling getDownloadableUrl function
+				let downloadableUrl = await filesHelpers.getDownloadableUrl(
+					payloadData,
+					bucketName,
+					cloudStorage,
+					typeof process.env.DOWNLOADABLE_URL_EXPIRY_IN_SECONDS === 'string'
```
Comment (VISHNUDAS-tunerlabs) [inline module/cloud-services/files/helper.js:128]: @Prajwal17Tunerlabs Here also, we can avoid the extra if condition check

File: module/files/helper.js [hunk 15]
```diff
@@ -1,126 +1,187 @@
...
-   * @name 								- preSignedUrls
-   * @param {Array} payloadData 		- payload for files data.
-   * @param {String} userId 			- Logged in user id.
-   * @param {String} templateId			- certificateTemplateId.
-   * @returns {Array} 					- consists of all signed urls files.
-   */
+	 * Get downloadable url
+	 * @method
+	 * @name                        - getDownloadableUrl
+	 * @param {filePath}            - File path.
```
Comment (VISHNUDAS-tunerlabs) [inline module/files/helper.js:23]: @Prajwal17Tunerlabs inside {} specify type. not variable name. that you do after

File: module/files/helper.js [hunk 16]
```diff
@@ -1,126 +1,187 @@
...
-				});
+			try {
+				// Override cloud storage provider name if provided explicitly.
+				if (storageName !== '') {
+					cloudStorage = storageName
+				}
+				// Override expireIn if env variable is present.
+				if (expireIn !== '') {
+					expireIn =
+						typeof process.env.DOWNLOADABLE_URL_EXPIRY_IN_SECONDS === 'string'
```
Comment (VISHNUDAS-tunerlabs) [inline module/files/helper.js:40]: @Prajwal17Tunerlabs remove this check

File: package.json [hunk 17]
```diff
@@ -31,6 +31,7 @@
 	},
 	"homepage": "https://github.com/ELEVATE-Project/project-service#readme",
 	"dependencies": {
+		"adm-zip": "^0.5.12",
```
Comment (VISHNUDAS-tunerlabs) [inline package.json:34]: @Prajwal17Tunerlabs Is this required

File: controllers/v1/cloud-services/files.js [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+     *  "request" : {
+     *  "5f72f9998925ec7c60f79a91": {
+     *     "files": ["uploadFile.jpg", "uploadFile2.jpg"]
+        }}
+      }
+     * @apiSampleRequest /project/v1/cloud-services/files/preSignedUrls
+     * @apiUse successBody
+     * @apiUse errorBody
+     * @apiParamExample {json} Response:
+     * * {
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/cloud-services/files.js:53]: @Prajwal17Tunerlabs response is correct right?

File: controllers/v1/programs.js [hunk 19]
```diff
@@ -570,17 +598,28 @@ module.exports = class  Programs extends Abstract{
...
     * @apiParamExample {json} Response:
+<<<<<<< HEAD
+    *{
+      "message": "You have joined this program successfully",
+      "status": 200,
+      "result": {
+        "_id" : "5ffbf8909259097d48017bbf"
+      }
+    }
+=======
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/programs.js:614]: @Prajwal17Tunerlabs remove these lines.

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
review - 2 - may - 2

File: controllers/v1/programs.js (hunk 19 above)
Comment (Prajwal17Tunerlabs) [reply controllers/v1/programs.js:614]: removed these unwanted characters

File: controllers/v1/cloud-services/files.js (hunk 18 above)
Comment (Prajwal17Tunerlabs) [reply controllers/v1/cloud-services/files.js:53]: yes, i checked the response object, its correct

File: package.json (hunk 17 above)
Comment (Prajwal17Tunerlabs) [reply package.json:34]: removed this package

File: controllers/v1/programs.js [hunk 20]
```diff
@@ -544,6 +544,34 @@ module.exports = class  Programs extends Abstract{
     * @apiHeader {String} X-App-Ver Appversion
     * @apiSampleRequest /project/v1/programs/join/5ffbf8909259097d48017bbf
     * {
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (Prajwal17Tunerlabs) [reply controllers/v1/programs.js:547]: removed these unwanted characters

File: controllers/v1/programs.js [hunk 21]
```diff
@@ -570,17 +598,28 @@ module.exports = class  Programs extends Abstract{
       },
       "requestForPIIConsent" : true
 }
+>>>>>>> project-template-creation-changes
```
[the remark this answers is not in the document]
Comment (Prajwal17Tunerlabs) [reply controllers/v1/programs.js:601]: removed these unwanted characters

File: controllers/v1/solutions.js (hunk 10 above)
Comment (Prajwal17Tunerlabs) [reply controllers/v1/solutions.js:284]: got it, removed these unwanted characters

File: generics/constants/common.js (hunk 11 above)
Comment (Prajwal17Tunerlabs) [reply generics/constants/common.js:78]: it wasnt used anywhere, i removed it

File: module/cloud-services/files/helper.js (hunk 13 above)
Comment (Prajwal17Tunerlabs) [reply module/cloud-services/files/helper.js:103]: removed the check

File: module/cloud-services/files/helper.js (hunk 14 above)
Comment (Prajwal17Tunerlabs) [reply module/cloud-services/files/helper.js:128]: removed the check

File: module/files/helper.js (hunk 15 above)
Comment (Prajwal17Tunerlabs) [reply module/files/helper.js:23]: got it, have specified the type

File: module/cloud-services/files/helper.js (hunk 12 above)
Comment (Prajwal17Tunerlabs) [reply module/cloud-services/files/helper.js:54]: removed the check

File: module/files/helper.js (hunk 16 above)
Comment (Prajwal17Tunerlabs) [reply module/files/helper.js:40]: removed the check

File: controllers/v1/solutions.js [hunk 22]
```diff
@@ -281,7 +281,11 @@ module.exports = class Solutions extends Abstract {
...
+    * @apiName add roles in scope
+=======
     * @apiName Add roles in scope
+>>>>>>> project-template-creation-changes
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/solutions.js:288]: @Prajwal17Tunerlabs  remove these . why we need

Comment (Prajwal17Tunerlabs) [reply controllers/v1/solutions.js:288]: removed those unwanted characters

File: envVariables.js [hunk 23]
```diff
@@ -61,127 +61,46 @@ let enviromentVariables = {
...
   },
-  "AZURE_ACCOUNT_KEY" : {
-    "message" : "Required Azure Account key",
-    "optional" : true,
-    "requiredIf" : {
-      "key": "CLOUD_STORAGE",
-      "operator": "EQUALS",
-      "value" : "AZURE"
-    }
+
```
Comment (VISHNUDAS-tunerlabs) [inline envVariables.js:81]: @Prajwal17Tunerlabs remove unwanted new lines please

Comment (Prajwal17Tunerlabs) [reply envVariables.js]: removed unwanted lines from the code

**Review by VISHNUDAS-tunerlabs — APPROVED** [review]

**Pull request merged**